rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,517 @@
+targets, we can establish a predefined histogram that would allow Couchbase to
+simply and inexpensively understand if the service level has likely changed.
+
+It's proposed that we establish a set of standardized histograms, specific to
+services and to the service level expected with the service using those
+parameters.  Counters for these histograms can be sent to Couchbase and both
+stored individually and aggregated.
+
+As an aggregated set of metrics, it would give Couchbase a rough view over the
+service level the user is experiencing.  Deviations in service level can be dug
+into and, in particular, we can check for correlation to Couchbase Capella
+management activities like upgrades.
+
+This list represents the services, buckets.
+
+* Data Service, Non-Durable
+  * Retrieval
+    * <1ms
+    * 1-10ms
+    * 10-100ms
+    * 100-500ms
+    * 500-1000ms
+    * 1s-2.5s
+    * \>2.5s
+  * Mutation
+    * <1ms
+    * 1-10ms
+    * 10-100ms
+    * 100-500ms
+    * 500-1000ms
+    * 1s-2.5s
+    * \>2.5s
+* Data Service, Durable
+  * Mutation
+    * 1-10ms
+    * 10-100ms
+    * 100-1000ms
+    * 1-2s
+    * 2s-5s
+    * 5s-10s
+    * \>10s
+* Query Service
+  * Query (all)
+    * 0-100ms
+    * 100-1000ms
+    * 1s-10s
+    * 10s-30s
+    * 30-75s
+    * \>75s
+* FTS Service
+  * Search (all)
+    * 0-100ms
+    * 100-1000ms
+    * 1s-10s
+    * 10s-30s
+    * 30-75s
+    * \>75s
+* Analytics Service
+  * Query (all)
+    * 0-100ms
+    * 100-1000ms
+    * 1s-10s
+    * 10s-30s
+    * 30-75s
+    * \>75s
+
+The Indexing Service and Eventing Services are not included as they are not
+network-available services for applications to make requests from.
+
+As with the Service Experience Counters above, it may be incorrect to try to
+compare between deployments.  For example, the query complexity may differ
+significantly or the data size and use of features like compression may vary
+significantly between two applications.
+
+# API and Implementation Details
+
+It should be possible to reuse existing Meter implementation, that according to
+RFC-67 (Extended SDK Observability) already builds histograms for request
+latencies.
+
+In addition to latencies, which would be used for reporting histograms, the
+default Meter should also maintain the following counters:
+
+| Metric | Description |
+| :---- | :---- |
+| sdk\_*{service}*\_r\_utimedout | Unambiguous timeout |
+| sdk\_*{service}*\_r\_atimedout | Ambiguous timeout |
+| sdk\_*{service}*\_r\_canceled | Canceled |
+| sdk\_*{service}*\_r\_total | Total number of operations |
+
+Where the service is a part of the visible counter name, and hostname and
+bucket must be stored and exported as a meta data.
+
+## App Telemetry Activation
+
+The collection and reporting of application telemetry should be activated by
+either of the following mechanisms:
+
+* presence of the `appTelemetry` property of the node in `nodexExt` entry of
```

Comment:
Ok so essentially we get an appTelemetry property that contains a wss:// address we need to connect to?

What assumptions are we allowed to make?  Are these safe:

1. We can use the same cert for the wss:// as we are using for the server.  E.g. we don't need to make provision for the user to configure it.

2. Can multiple nodes in nodesExt report appTelemetry?  Will it always be the same or do we need to handle having multiple connections.  Update: I think the below suggests there can be multiple hostnames, but we should only connect to one, and round-robin to the next if we need to re-establish?

3. We can get a config that no longer contains the hostname we have connected to, but this needs no special handling.  (Since if a node has been rebalanced out, our connection should have been severed, and we will already be attempting connection to another hostname.)

4. And same applies if no nodes have appTelemetry - e.g. that implicitly disables the feature with no special handling needing.

Reply:
N.b. this was further clarified in a recent meeting.  The idea is the server sends some indication that (an? all?) existing connection(s) to ns-server should be upgraded to websocket - rather than a new connection being created.

And hence we can use the existing server cert - which answers Q1.

Reply:
Configuration property is not specified yet. It is up to the Server team to define name and semantics of the property.

Reply:
We discussed further in a meeting.  Results are:

* The feature doesn't seem to be particularly specced on the server side yet, and so we don't know the answer to the questions above.
* We should assume that the server team will want to be able to disable the feature on their side dynamically, but we do not yet know what that will look like.  My assumption: they will simply stop scraping us for metrics.  The feature may or may not still appear in the config when it is disabled.
* Per Sergey: regardless of what the server does, we should always collect the metrics.  Partly so we can provide them immediately if the server does enable it again.  Plus it simplifies implementation.

Reply:
In the current [server spec][1] it says that the property will have full URL, but I'd like to convince them to use just path, and let SDK to infer the host and scheme (which we really do not need). It will also make it easier for server too, as they do not know whether the SDK is going to use alternative addresses or default.

[1]: https://docs.google.com/document/d/1ZbqiooiyQL2TmMIqtASO2tR3U1FrrVky3wXYN-JknHo/edit